modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/LocalDatabases.java

```diff
@@ -0,0 +1,208 @@
+                DatabaseReaderLazyLoader existing = databases.put(databaseFileName, loader);
+                if (existing != null) {
+                    existing.close();
+                    int numEntriesEvicted = cache.purgeCacheEntriesForDatabase(file);
+                    LOGGER.info("evicted [{}] entries from cache after reloading database [{}]", numEntriesEvicted, file);
+                }
+            } else {
+                LOGGER.info("database file removed [{}], close database...", file);
+                DatabaseReaderLazyLoader existing = databases.remove(databaseFileName);
+                assert existing != null;
+                existing.close();
+                int numEntriesEvicted = cache.purgeCacheEntriesForDatabase(file);
+                LOGGER.info("evicted [{}] entries from cache after reloading database [{}]", numEntriesEvicted, file);
+            }
+            this.configDatabases = Map.copyOf(databases);
+        } catch (Exception e) {
+            LOGGER.error((Supplier<?>) () -> new ParameterizedMessage("failed to update database [{}]", databaseFileName), e);
+        }
+    }
+
+    Map<String, DatabaseReaderLazyLoader> initDefaultDatabases(Path geoipModuleDir) {
+        Map<String, DatabaseReaderLazyLoader> databases = new HashMap<>();
+
+        for (String filename : DEFAULT_DATABASE_FILENAMES) {
+            Path source = geoipModuleDir.resolve(filename);
+            assert Files.exists(source);
+            String databaseFileName = source.getFileName().toString();
+            DatabaseReaderLazyLoader loader = new DatabaseReaderLazyLoader(cache, source);
+            databases.put(databaseFileName, loader);
+        }
+
+        return Map.copyOf(databases);
+    }
+
+    Map<String, DatabaseReaderLazyLoader> initConfigDatabases(Path geoipConfigDir) throws IOException {
+        Map<String, DatabaseReaderLazyLoader> databases = new HashMap<>();
+
+        if (geoipConfigDir != null && Files.exists(geoipConfigDir)) {
+            try (Stream<Path> databaseFiles = Files.list(geoipConfigDir)) {
+                PathMatcher pathMatcher = geoipConfigDir.getFileSystem().getPathMatcher("glob:**.mmdb");
+                // Use iterator instead of forEach otherwise IOException needs to be caught twice...
+                Iterator<Path> iterator = databaseFiles.iterator();
+                while (iterator.hasNext()) {
+                    Path databasePath = iterator.next();
+                    if (Files.isRegularFile(databasePath) && pathMatcher.matches(databasePath)) {
+                        assert Files.exists(databasePath);
+                        String databaseFileName = databasePath.getFileName().toString();
+                        DatabaseReaderLazyLoader loader = new DatabaseReaderLazyLoader(cache, databasePath);
+                        databases.put(databaseFileName, loader);
+                    }
+                }
+            }
+        }
+
+        return Map.copyOf(databases);
```

Comment:
Any reason for creating copy here? Collections.unmodifiableMap(databases) should suffice I think

Reply:
I will change this, see the previous comment.